util.go

```diff
@@ -11,3 +16,47 @@ var KValue = 20
 // Deprecated: This value is still the default, but changing this parameter no longer does anything.
 // Alpha is the concurrency factor for asynchronous requests.
 var AlphaValue = 3
+
+// RoutabilityBasedModeSwitching uses LocalRoutabilty events sent over the DHT's EventBus to trigger switching between
+// client and server modes.
+func RoutabilityBasedModeSwitching(dht *IpfsDHT) {
+	const modeBufSz = 256
+
+	evts := []interface{}{
+		&event.EvtLocalRoutabilityPublic{},
+		&event.EvtLocalRoutabilityPrivate{},
+		&event.EvtLocalRoutabilityUnknown{},
+	}
+
+	sub, err := dht.Host().EventBus().Subscribe(evts, eventbus.BufSize(modeBufSz))
+	if err != nil {
+		logger.Errorf("kaddht not subscribed to local routability events; dynamic mode switching will not work; err: %s", err)
+	}
+	defer sub.Close()
+
+	for {
+		select {
+		case ev := <-sub.Out():
+			var target DHTMode
+
+			switch ev.(type) {
+			case event.EvtLocalRoutabilityPrivate, event.EvtLocalRoutabilityUnknown:
+				target = ModeClient
+			case event.EvtLocalRoutabilityPublic:
+				target = ModeServer
+			}
+
+			logger.Infof("processed event %T; performing kaddht mode switch", ev)
+
+			err := dht.SetMode(target)
+			// NOTE: the mode will be printed out as a decimal.
+			if err == nil {
+				logger.Infof("switched DHT mode successfully; new mode: %d", target)
+			} else {
+				logger.Warningf("switching DHT mode failed; new mode: %d, err: %s", target, err)
+			}
+		case <-dht.Context().Done():
```

Comment:
Should we be using `Context.Done` or `Process.Closing` here?

Reply:
They should be equivalent.

Reply:
@Stebalien just wanted to double check given your comment on 96d6e8173b2414c1a5c2b7ab52a1b7dba3b81780

Reply:
Hm. Nevermind, you're right. We _should_ be canceling that context but we aren't. Let's use the process for now. I've filed https://github.com/libp2p/go-libp2p-kad-dht/issues/476.

Reply:
Fix for #476 is at https://github.com/libp2p/go-libp2p-kad-dht/pull/477.